### ISSUE TITLE
Supervisor should implement "happy eyeballs"

### ISSUE THREAD

**Issue by zwhitchcox** [balena-os/balena-supervisor#1787]
UPDATE

This can now be fixed, by simply adding:

```ts
import 'happy-eyeballs/eye-patch'
```

at the entrypoint of the nodejs app

-----------------

There was an issue where the supervisor wasn't sending a heartbeat. This was because the network had dropped IPv4 support and was only serving IPv6.

Furthermore, the network wasn't sending an ICMP Host Unreachable packet, because when the request was failing, a traceroute to `api.balena-cloud.com` would get to the `192.168.1.1` and then just hang.

This caused [`getaddrinfo`](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) to return the wrong address. It was still returning the IPv4 address, because it had never gotten any indication that that address no longer worked.

According to [RFC 3484](https://www.ietf.org/rfc/rfc3484.txt), rule 1 is `Avoid unusable destinations`, and the kernel was never informed the destination was unusable.

But it also says

> Well-behaved applications SHOULD iterate through the list of addresses returned from getaddrinfo() until they find a working address.

`curl` responses were working because it tried both IPv6 and IPv4 in parallel:

```
root@fc09813:~# curl -vv https://api.balena-cloud.com/ping
*   Trying 54.164.146.95:443...
*   Trying fd00:0:b:33::3693:e3ac:443...
```

Then, it dropped the connection if one wasn't working properly, preferring IPv6 if both were working.

But that functionality is not built into the supervisor. It just tries the first IP address returns by `getaddrinfo`, and fails if it can't connect. I checked when it wasn't getting a heart beat, and it returned the IPv4 address still, even when IPv6 only was working.

So, that means the supervisor devs SHOULD try different IP addresses if the first one fails. That will involve creating our own custom [`Agent`](https://nodejs.org/api/http.html#http_new_agent_options) that extends the [`createConnection` method](https://nodejs.org/api/http.html#http_agent_createconnection_options_callback) OR just do like is [currently done with the dns.lookup function](https://github.com/balena-os/balena-supervisor/blob/master/src/app.ts#L32) and monkey patch the [`net.createConnection` globally](https://nodejs.org/api/net.html#net_net_createconnection), which I don't think is a terrible idea, in fact this functionality should probably be built into Node really. Maybe I should open an issue in nodejs/node.

But really, the ideal solution would probably be extending the `Agent` class and use that over all http connections to be more explicit and avoid future conflicts, although, the supervisor is so well maintained and only used by us and version controlled, it probably doesn't really matter.

The algorithm should probably try both IPv4 and IPv6 connections in parallel, just like curl and drop the non-responsive one.

This wouldn't be the final code, but it could be something like:

```js
import net from 'net';

const {createConnection} = net;
net.createConnection = async (path, cb) => {
  const addresses = await dnsPromises.lookup(path, {verbatim: true});
  const ipv6 = addresses.find(({family}) => family === 6);
  const ipv4 = addresses.find(({family}) => family === 4);
  try {
    cb(null, await Promise.race([
      new Promise((res) => createConnection(ipv6, res),
      new Promise((res) => createConnection(ipv4, res),
    ]));
  } catch (err) {
    cb(err);
  }

}
```

**Comment by zwhitchcox** [comment]
I later learned that this is called [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs), and it is enabled in the browser and on mobile by default.

**Comment by zwhitchcox** [comment]
There is now a fix for this, simply importing `happy-eyeballs/eye-patch` at the time of the entrypoint:

```ts
import 'happy-eyeballs/eye-patch'
```

**Comment by 20k-ultra** [comment]
closed in #1805
